[Low Jia Hao] iP

LowJiaHao99 opened the pull request:
DukePro

 > “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Always having too many tasks to keep track of? DukePro is here to help you 😉 
It is :
- text-based
- easy to learn
- ***SUPER FAST*** to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2122s2.github.io/website/schedule/week4/project.html)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 

And it is **FREE**! 👍 

Features:
- [x] Managing tasks
- [x] Managing deadlines 
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

LowJiaHao99 committed:
added echo,exit and greet function

LowJiaHao99 committed:
Unable to keep track of commands

Keep track of commands so users can see past inputs

Added array in Command class to keep track of past input and list when called

list and bye command are not added in tracking of past commands

LowJiaHao99 committed:
created a Task class

majority of commands is a form of task and they have similar behavior(e.g. isDone and have a description)

To allow for future polymorphisms where there are different types of tasks that is being tracked

LowJiaHao99 committed:
commands now are handled directly by Duke instead of the Commands class

commands can be directly processed by Duke class

LowJiaHao99 committed:
Removed command class

commands are now processed by Duke class

LowJiaHao99 committed:
Added text to match expected output when calling list.
Changed to print "added input" where input is the whole input and not only the first word.
Simplified some logic in the program

LowJiaHao99 committed:
added Task class

To allow future classes to inherit from

added way to identify type of task and indicate whether is completed

to allow printing of output just by calling identity method

LowJiaHao99 committed:
added Deadline class

Deadline inherits from Task. Deadline is a task with extra fields(e.g. the due date of said task)

Overrided identify so as differentiate deadline from tasks

LowJiaHao99 committed:
Added ToDo class

ToDo is a form of task. Similar fields and behavior

Overrided identify so as to  differentiate ToDo from Task

LowJiaHao99 committed:
Added Event class

Event class is a Task thus it inherits from Task
Override identify class to differentiate Task from Events.
Extra field of deadline to indicate the due date of event

LowJiaHao99 committed:
Added extensions to commands so as to accomodate new classes(i.e. Task, ToDo, Event and Deadline)

processedCommand is further processed to obtain deadline and description which are used to create Event and/or Deadline instances
Logic of printing whether task is marked or unmarked is now handle by the individual classes(Event,Deadline,ToDo,Task)

LowJiaHao99 committed:
Modify expected.txt and input.txt for testing of level-4

LowJiaHao99 committed:
Add DukeException class

DukeException prints message whenever called

LowJiaHao99 committed:
Modify DukeException

DukeException prints error message

LowJiaHao99 committed:
Modify Duke class

Change print statements to easily identify what is printed
Add exception handling for commands where exception may occur due to invalid input

LowJiaHao99 committed:
Modify Duke class

Handle "hi" command
Invalid message for invalid commands

LowJiaHao99 committed:
Change task array to task arraylist
remove numOfTask and use arraylist.size() to get number of task

LowJiaHao99 committed:
Added delete command
Modify presentation of print statement for unmark command

LowJiaHao99 committed:
Modify input.txt and expected.txt to match new delete command

LowJiaHao99 committed:
Add delete command
Modify string output by removing unecessary whitespace

LowJiaHao99 committed:
Remove whitespaces at the end

LowJiaHao99 committed:
Add commands as enum

Now string that are associated with the command is printed via the enum Commands
Change print statements in Duke class to reflect enums

LowJiaHao99 committed:
Modify coding standard for switch statements

LowJiaHao99 committed:
add save feature

LowJiaHao99 committed:
Update deadlines to be kept as a LocalDate class instead of a string

LowJiaHao99 committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Duke.java

LowJiaHao99 committed:
Remove uneccessary import statements

LowJiaHao99 committed:
Refactor code

Parser parses user input and return parsedCommand to Duke
Storage deals with interaction with the storage file
TaskList contains data in storage file
UI deals with output to users
Duke coordinate interaction with all the Storage,Tasklist,Parser and Ui

LowJiaHao99 committed:
Refactor code.

Parser parses user input and return parsedCommand to Duke
Storage deals with interaction with the storage file
TaskList contains data in storage file
UI deals with output to users
Duke coordinate interaction with all the Storage,Tasklist,Parser and Ui

LowJiaHao99 committed:
Shift all classes to DukeClasses package

LowJiaHao99 committed:
Shift all classes to DukeClasses package

LowJiaHao99 committed:
Update deadlines to be kept as a LocalDate instead of a String

LowJiaHao99 committed:
Merge branch 'branch-Level-7' into branch-Level-8

# Conflicts:
#	src/main/java/Duke.java

LowJiaHao99 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

LowJiaHao99 committed:
Remove duplicate files

LowJiaHao99 committed:
Change string output of Event identity method

LowJiaHao99 committed:
Add JUnit tests

LowJiaHao99 committed:
Add StorageTest class

LowJiaHao99 committed:
Modify Duke, Storage and Ui classes

Add calls to close Scanner
Change storageFilePath to a static String
Modify method to throw DukeException

LowJiaHao99 committed:
Fix bug for delete command

LowJiaHao99 committed:
Modify Event,Duke and Deadline classes

LowJiaHao99 committed:
Remove Deadline and Event

LowJiaHao99 committed:
Disable tracking of a workspace.xml

LowJiaHao99 committed:
Update Ui and TaskList to run modified deleted command

LowJiaHao99 committed:
Add new files into .gitignore

LowJiaHao99 committed:
Remove revision control for Duke.jar

LowJiaHao99 committed:
Add data.txt file into .gitignore

LowJiaHao99 committed:
Add JavaDocs for classes

LowJiaHao99 committed:
Modify code and pacakges to fit coding standards

LowJiaHao99 committed:
Add find feature

LowJiaHao99 committed:
Change output for find feature

LowJiaHao99 committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/dukeclasses/Deadline.java
#	src/main/java/dukeclasses/Duke.java
#	src/main/java/dukeclasses/DukeException.java
#	src/main/java/dukeclasses/Event.java
#	src/main/java/dukeclasses/ParsedCommand.java
#	src/main/java/dukeclasses/Parser.java
#	src/main/java/dukeclasses/Storage.java
#	src/main/java/dukeclasses/Task.java
#	src/main/java/dukeclasses/TaskList.java
#	src/main/java/dukeclasses/Ui.java

LowJiaHao99 committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

# Conflicts:
#	src/main/java/dukeClasses/Ui.java
#	src/main/java/dukeclasses/Commands.java

LowJiaHao99 committed:
Modify Commands and Ui to fix find and delete feature

LowJiaHao99 committed:
Merge branch 'branch-Level-9'

LowJiaHao99 committed:
Merge branch 'add-gradle-support' of https://github.com/LowJiaHao99/ip

LowJiaHao99 committed:
Implement gradle

LowJiaHao99 committed:
Update .gitignore

LowJiaHao99 committed:
Merge branch 'A-Gradle'

AD-NAP commented on src/main/java/dukeclasses/ParsedCommand.java:
```diff
@@ -0,0 +1,105 @@
+    private String command;
+    private Integer index;
+    private LocalDate dueDate;
+    private String task;
```
Perhaps this `String task` could be named as `taskName` to be more specific?

AD-NAP commented on src/main/java/dukeclasses/ParsedCommand.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * Constructor for ParsedCommand for commands that only need String to be instantiated.
+     *
+     * @param command String , usually the command, that is used to instantiated Tasks.
```
Should the comma after String appear? Perhaps it could be rephrased. Using 'usually' implies that there is an occasion that command is not used to instantiate Tasks. If so, what is that special cases? I notice the same issue in other places too. Worth thinking about it 😅 
 
```suggestion
     * @param command String the action requested to instantiate Tasks. 
```

AD-NAP commented on src/main/java/dukeclasses/Parser.java:
```diff
@@ -0,0 +1,67 @@
+     *
+     * @param userCommand String that the user inputs.
+     * @return ParsedCommand that denotes the command and other relevant information if the user input is valid.
+     * @throws DukeException If user input is invalid.
```
Maybe the if does not need to be capitalise?

AD-NAP commented on src/main/java/dukeclasses/Storage.java:
```diff
@@ -0,0 +1,150 @@
+package dukeclasses;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+import java.util.ArrayList;
+import java.util.Scanner;
```
Nice to see the import statements are grouped as mentioned in the coding standard 😄

AD-NAP commented on src/main/java/dukeclasses/Storage.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Loads data of the storage file into an ArrayList</Task>. Usually called at the start of Duke.
```
This might violate the coding standard. It mentioned 
> Write the first sentence as a short summary of the method, as Javadoc automatically places it in the method summary table (and index).

Perhaps the second sentence can be underneath the first

AD-NAP commented on src/main/java/dukeclasses/Task.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return Boolean that indicate whether task is done or not.
+     */
+    public boolean getIsDone() {
```
The java coding standard mentions 
>  Boolean variables/methods should be named to sound like booleans

Could the method also be named to sound like booleans? Perhaps isDone() ? Even though this is the "same" as the variable, you will not access the variable directly as it is set to private. So the IDE should still be able to help to recommend the isDone() method whenever you try to access it and not the variable.

AD-NAP commented on src/main/java/dukeclasses/Task.java:
```diff
@@ -0,0 +1,62 @@
+        return description;
+    }
+
+    /**
+     * Returns description of the current status of the task(its description and whether it is marked or not).
+     * Identity used in taskList.
+     *
+     * @return String that identify the task
+     */
+    public String identify() {
+        if (isDone) {
+            return String.format("[X] %s", description);
+        } else {
+            return String.format("[ ] %s", description);
+        }
+    }
```
Perhaps the word identify does not suit the case here? Identify sounds like it's meant to identify if this Task is a ToDo or Deadline etc. Since it is returning a string with full details of isDone and the description, perhaps the method could be called getFullDetails() ? Or possibly override the toString() method as it's not utilised yet

AD-NAP commented on src/main/java/dukeclasses/Ui.java:
```diff
@@ -0,0 +1,135 @@
+package dukeclasses;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
```
Perhaps add a blank link between io and util imports, to be consistent with the other files

AD-NAP reviewed (commented):
Overall, good job on following the java coding standard closely. Very minor mistakes that I could spot. 
One thing I noticed is how you format your codes. You tend to have a blank line to separate variable declaration and if else or switch case statements. However, this might lead to instances where it does not follow this format (I notice it mainly in Duke.java)
Nonetheless, it does make it look easier to read, but personally I do not do that (I might try to) 😄

LowJiaHao99 committed:
Create GUI for Duke

LowJiaHao99 committed:
Improve feature of GUI and add new background for GUI

LowJiaHao99 committed:
Merge branch 'branch-Level-10'

LowJiaHao99 committed:
Fixes bug for delete command

talkintomato commented on src/main/Duke.java:
```diff
@@ -0,0 +1,165 @@
+                }
+
+                //Input processing
+                String[] inputForEventConstructor = processedCommand[1].split("/by", 2);
+                inputForEventConstructor[0] = inputForEventConstructor[0].trim();
+                LocalDate eventDate = null;
+                try {
+                    eventDate = LocalDate.parse(inputForEventConstructor[1].trim());
+                } catch (DateTimeParseException e) {
+                    new DukeException();
+                    break;
+                }
+
+                //Create event
+                Event event = new Event(inputForEventConstructor[0], eventDate);
+                System.out.println(String.format(
+                    "%s      %s", Commands.ADD.toString(), event.identify()));
+                tasks.add(event);
+                System.out.println(String.format("    Now you have %d tasks in the list.", tasks.size()));
+                break;
+            case "delete":
+                //Exception catching
+                if (processedCommand.length <= 1) {
+                    new DukeException();
+                    break;
+                }
+                try {
+                    taskIndex = Integer.parseInt(processedCommand[1]) - 1;
+                } catch (NumberFormatException e) {
+                    new DukeException();
+                    break;
+                }
+
+                Task deletedTask = tasks.remove((int)taskIndex);
+                System.out.println(String.format(
+                    "%s      %s\n    Now you have %d tasks in the list",
+                    Commands.DELETE.toString(), deletedTask.identify(), tasks.size()));
+                break;
+            default:
+                new DukeException();
```
can try abstracting this more (make it more OOP) through the use of a commands and class!

talkintomato commented on src/main/java/dukeclasses/Duke.java:
```diff
@@ -0,0 +1,314 @@
+        }
+
+        switch (parsedCommand.getCommand()) {
+        case "hi":
```
think you may be missing an indent on your case switch!

talkintomato reviewed (commented):
Good job overall sticking with the coding standards! Found your code easy to understand

LowJiaHao99 committed:
Fix bug for Parser class where negative index will crash the client

LowJiaHao99 committed:
Refactor classes in dukeclasses package

classes in dukeclasses package does not follow code quality standards.

Let's pull up and group behaviors as new methods in the class.

This is to ensure code follow code quality standards by enforcing single
level of abstraction and prevent deep nesting in a method to enhance
readability.

LowJiaHao99 committed:
Update test classes to most recent output

Test classes are not up to date.

This ensures future testing can be done when new features are added.

Let's edit the expected cases in the test classes to fit the new output.

Actual output contains a new line character at the end of a toString
method of a task. The change modifies expected to meet the new change
in the actual output.

LowJiaHao99 committed:
Modify the Ui class

LowJiaHao99 committed:
Refactor Parser and Storage classes

LowJiaHao99 committed:
Add assert statements

Classes contain assumptions.

Assumptions does not always hold true and when it does not, it causes
unexpected behavior.

Let's add assert statements in the code when there is an assumption in
all the classes.

Assert statements crashes the program if the assumption fails,thus
prevents unexpected behavior.

LowJiaHao99 committed:
Remove unused variable in Duke and Storage class

LowJiaHao99 committed:
Merge pull request #1 from LowJiaHao99/A-CodeQuality

Improve code quality

LowJiaHao99 committed:
Merge branch 'master' into A-Assertions

# Conflicts:
#	src/main/java/dukeclasses/Duke.java
#	src/main/java/dukeclasses/Parser.java
#	src/main/java/dukeclasses/Storage.java

LowJiaHao99 committed:
Merge pull request #2 from LowJiaHao99/A-Assertions

Add assert statements

LowJiaHao99 committed:
Fix bug for Parser that affects creating deadlines and events

LowJiaHao99 committed:
Add extension to manage recurring task and update test classes

LowJiaHao99 committed:
Merge branch 'B-RecurringTask'

LowJiaHao99 committed:
Improve GUI, refactor code and update testclasses

LowJiaHao99 committed:
Fix bug for recurring tasks

LowJiaHao99 committed:
Add Ui.png to docs folder

LowJiaHao99 committed:
Add image credits

LowJiaHao99 committed:
Modify messages to print in the Ui and update User Guide

LowJiaHao99 committed:
Change Ui.png to reflect changes made to Ui class

LowJiaHao99 committed:
Set theme jekyll-theme-cayman

LowJiaHao99 committed:
Modify User Guide

LowJiaHao99 committed:
Merge branch 'master' of https://github.com/LowJiaHao99/ip

LowJiaHao99 committed:
Refactor code to improve code quality

LowJiaHao99 committed:
Update User Guide and method descriptions

LowJiaHao99 committed:
Update README.md
